pkg/provenance/provenance.go

```diff
@@ -164,105 +170,335 @@ func (o ObjectLineage) SpecHistory() string {
+	specs := make([]Spec, 0)
+	for _, version := range s {
+		specs = append(specs, o[version])
+	}
+	allQueryPairs := make([][]string, 0)
+	//this section is storing the mappings from the query, into the allQueryPairs object
+	//these are the queries, each element must be satisfied somewhere in the spec for the bisect to succeed
+	for key, value := range argMap {
+		attributeValueSlice := make([]string, 2)
+		if strings.Contains(key, "field") {
+			//find associated value in argMap
+			fieldNum, err := strconv.Atoi(key[5:])
+			if err != nil {
+				return fmt.Sprintf("Failure, could not convert %s. Invalid Query parameters.", key)
+			}
+			//find assocaited value1
+			valueOfKey, ok := argMap["value"+strconv.Itoa(fieldNum)]
+			if !ok {
+				return fmt.Sprintf("Could not find an associated value for field: %s", key)
+			}
+			attributeValueSlice[0] = value
+			attributeValueSlice[1] = valueOfKey
+			allQueryPairs = append(allQueryPairs, attributeValueSlice)
+		}
+	}
+	// fmt.Printf("attributeValuePairs%s\n", allQueryPairs)
+
+	//this section is to build the attributeRelationships from the Query
+	//will use this to ensure that fields belonging to the same top-level
+	//attribute, will be treated as a joint query. So you can't just
+	//ask if username ever is daniel and password is ever 223843, because
+	//it could find that in different parts of the spec. they both must be satisfied in the same map object
+	andGate := make([]bool, len(allQueryPairs))
+	mapRelationships := make(map[string][][]string, 0) // a map from top level attribute to array of pairs (represented as 2 len array)
+	for _, spec := range specs {
+		for _, pair := range allQueryPairs {
+			for mkey, mvalue := range spec.AttributeToData {
+
+				qkey := pair[0] //query key
+				//each qkey qval has to be satisfied
+				vSliceMap, ok3 := mvalue.([]map[string]string)
+				if ok3 {
+					for _, mymap := range vSliceMap {
+						for okey, _ := range mymap {
+							if qkey == okey {
+								pairs, ok := mapRelationships[mkey]
+								if ok {
+									mapRelationships[mkey] = append(mapRelationships[mkey], pair)
+								} else {
+									pairs = make([][]string, 0)
+									pairs = append(pairs, pair)
+									mapRelationships[mkey] = pairs
+								}
+
+							}
+						}
+					}
+				}
+			}
+		}
+		index := 0
+		for _, pair := range allQueryPairs {
+			qkey := pair[0]
+			qval := pair[1]
+			satisfied := false
+			for mkey, mvalue := range spec.AttributeToData {
+				vString, ok1 := mvalue.(string)
+				if ok1 { //checking string
+					if qkey == mkey && qval == vString {
+						satisfied = true
+						break
+					}
+				}
+				vStringSlice, ok2 := mvalue.([]string)
+				if ok2 { //p
+					for _, str := range vStringSlice {
+						if qkey == mkey && qval == str {
+							satisfied = true
+							break
+						}
+					}
+				}
+				vSliceMap, ok3 := mvalue.([]map[string]string)
+				if ok3 {
+					fmt.Println("c")
```

Comment:
Lines 297 - 347 is where the main logic of searching is happening, correct?

I am wondering if this can be simplified so that it becomes easy to follow. 

When I was envisioning how we might end up doing bisect, I was thinking there will be two cases: 1) when the field is top-level field such as 'databases'. 2) when the field is not top-level such as 'username' and 'password' OR 1) when the field is a simple field ('databases'). 2) when the field is a composite filed ('username' and 'password').

The first option is closer to the structure of the Spec whereas second option is focusing more on the structure of the fields rather than the Spec.

I think second option is probably what you have implemented.

For simplification, I think we should consider breaking out the two cases into separate functions, say:
func handleSimpleFields()

func handleCompositeFields()

That way it will be easier to follow the logic.

Reply:
I see, Yes I made methods like you said. To clarify a little bc I think there is one small misunderstanding, the logic for databases top level attribute IS separate from the logic for username/password, that code is above it (where I type assert []string). Lines 297-347 are only for the composite query!

right now I have 3 separate cases with different logic that follows each of these lines
vString, ok1 := mvalue.(string)
vStringSlice, ok2 := mvalue.([]string)
vSliceMapSlice, ok3 := mvalue.([]map[string]string)

Like you said, the last line, after ok3, starting on 297, is how I handle the case where the field is not a top-level like username and password, the underlying data is the slice of maps[these maps contain attributes]. So I think I can only put this one in the handleCompositeFields, the reason there is an if/else in that method is because sometimes someone may just query whether there exists a username="pallavi" but the query is much simpler as there is no interdependence on password .  So, I don't need to use the mapRelationships object to validate it, hence the else

Anyway, I added 3 methods, similar to how you suggested but I am afraid that there has to be three since my logic is broken up into 3 type assertions, of different types.

Let me know if this looks ok, I can make further changes.